### ISSUE TITLE
Different results from Luong

### ISSUE THREAD

**Issue by tulongthienvu** [OpenNMT/OpenNMT-py#760]
Hi everyone,

I am trying to replicate the results of Luong's paper https://arxiv.org/pdf/1508.04025.pdf. I set up settings as Luong's settings and code as https://github.com/lmthang/nmt.matlab. The dataset I used is the same with data provided in https://nlp.stanford.edu/projects/nmt/ (same vocabs, same train, valid, test datasets).

My problem is my results always better than Luong's results (BLEU, Perplexity) and results of Global Attention with general product is strange. According to the paper, with Global Attention, dot product is better than general product. But in my case, general product is always better than dot product even (maybe) better than my implement of Local Attention (local-p with general product) (because I just trained 10 epochs of global general; all models were trained with 12 epochs). I translated all models with beam size 12. You can see detailed results in the table in this link: https://drive.google.com/file/d/1UtGVbMDp68t9SQdCq65S15rXMxDqFfLY/view

Up to now, I just have figured out that I have a small difference with Luong's models that is (maybe) Luong used bi-LSTM for both encoder and decoder and OpenNMT-py is bi-LSTM for encoder and uni-LSTM for decoder. But I think this is not a big difference that causes to my problem.

Anyone had experiences with OpenNMT-py please to help me explain why that happened when I use OpenNMT-py to replicate Luong's paper results. My mentor asked me to explain but I cannot find out the reasons.

Thank you all very much.

**Comment by srush** [comment]
Our model is based on Luong's paper, but there are some slight differences with his code, in particular in initialization and optimization. I am not sure you will be able to replicate his results exactly. Sorry about that.

**Comment by tulongthienvu** [comment]
Can you explain more detailed information about the differences?
I used the same initialization with Luong's paper (all parameters are initialized in range [-0.1;0.1]), same vanilla SGD with learning rate schedule.
My friend told me that is about differences between Matlab and PyTorch. Luong implemented all models, gradient calculations by hand and but in OpenNMT-py it is powered by NVIDIA and PyTorch's autograd. But I don't think it is the main reason.

Thank you very much!

**Comment by vince62s** [comment]
see #668
